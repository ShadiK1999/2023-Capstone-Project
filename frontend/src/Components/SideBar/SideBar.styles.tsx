import { style } from 'typestyle';

export const Container = style({
  backgroundColor: '#7C96AB',
  borderRadius: '6px',
  padding: '22px 14px',
  maxHeight: '100%',
  color: 'white',
});

export const InnerContainer = style({
  height: '100%',
  display: 'flex',
  flexDirection: 'column',
  $nest: {
    '& > * + *': {
      marginTop: '16px',
    },
    '&::-webkit-scrollbar': {
      display: 'none',
    },
  },
  overflowY: 'scroll',
  overflowX: 'hidden',
  scrollbarWidth: 'none',
});

export const Panel = style({
  backgroundColor: '#BFCCB5',
  borderRadius: '6px',
  width: '100%',
  display: 'flex',
  flexDirection: 'column',
  $nest: {
    '& > * + *': {
      marginTop: '2px',
    },
  },
  padding: '10px 16px',
  boxShadow: '2px 5px #888888',
});

export const Title = style({
  fontSize: '20px',
  fontWeight: 600,
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
  overflow: 'hidden',
});

export const Address = style({
  fontSize: '20px',
  fontWeight: 400,
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
  overflow: 'hidden',
});

export const SubtitlesPanel = style({
  display: 'grid',
  marginTop: '22px !important',
  gridTemplateColumns: '150px repeat(2, minmax(0, 1fr))',
  $nest: {
    '& > * + *': {
      marginLeft: '10px',
    },
  },
});

export const Subtitles = style({
  display: 'flex',
  $nest: {
    '&>img': {
      height: '24px',
      width: '24px',
    },
    '&>p': {
      fontSize: '15px',
      fontWeight: 300,
      marginLeft: '2px',
    },
  },
});
