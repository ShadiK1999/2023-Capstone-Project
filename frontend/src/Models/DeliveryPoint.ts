export interface TrustedPoint extends DeliveryPoint {
  id: string;
}

export interface DeliveryPoint {
  id: string | undefined;
  location: google.maps.LatLngLiteral;
  hours: OpeningHours;
  address: string;
}

export interface DisplayPoint extends DeliveryPoint {
  distanceText: string;
  timeText: string;
}

export type Time = {
  hour: number;
  minute: number;
};
type OpeningHour = {
  openingTime: Time;
  closingTime: Time;
};
export type OpeningHours = Record<number, OpeningHour>;
