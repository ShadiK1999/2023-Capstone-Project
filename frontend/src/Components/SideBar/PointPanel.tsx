import { FC } from 'react';
import * as styles from './SideBar.styles';
import Walking from './Assets/Walking.svg';
import Target from './Assets/Target.svg';
import Clock from './Assets/Clock.svg';
import { DisplayPoint } from '../../Models/DeliveryPoint';
import { printTime } from '../../Services/LocationsApiClient';

export type ComponentProp = {
  point: DisplayPoint;
};

const PointPanel: FC<ComponentProp> = ({ point: { address, distanceText, timeText, hours } }) => {
  const date = new Date().getDay();
  const time = hours[date];
  return (
    <div className={styles.Panel}>
      {/* TODO: change into actual name once we get it from BE*/}
      <p className={styles.Title}>Location</p>
      <p className={styles.Address}>{address}</p>
      <div className={styles.SubtitlesPanel}>
        <SubtitleDiv
          svg={Clock}
          text={time ? `${printTime(time.openingTime)} - ${printTime(time.closingTime)}` : 'Closed'}
        />
        <SubtitleDiv svg={Walking} text={timeText} />
        <SubtitleDiv svg={Target} text={distanceText} />
      </div>
    </div>
  );
};

const SubtitleDiv: FC<{ svg: string; text: string }> = ({ svg, text }) => {
  return (
    <div className={styles.Subtitles}>
      <img src={svg} />
      <p>{text}</p>
    </div>
  );
};

export { PointPanel };
