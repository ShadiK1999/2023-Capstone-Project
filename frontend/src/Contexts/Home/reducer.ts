import { HomeState, Filter } from './state';
import { IsochroneRingsOutput } from '../../Services/Isochrone';
import { produce } from 'immer';
import { DisplayPoint } from '../../Models/DeliveryPoint';

export type HomeAction =
  | {
      type: 'ISOCHRONE_READY';
      payload: IsochroneRingsOutput;
    }
  | {
      type: 'ADDRESS_CHANGED';
      payload: google.maps.LatLngLiteral;
    }
  | {
      type: 'FETCH_DELIVERY_POINTS_SUCCESS';
      payload: DisplayPoint[];
    }
  | {
      type: 'START_LOADING';
    }
  | {
      type: 'LOAD_SUCCESSFUL';
    }
  | {
      type: 'FILTER_INPUT_CHANGED';
      payload: Filter;
    };

export const homeReducer = produce((draft: HomeState, action: HomeAction) => {
  switch (action.type) {
    case 'ISOCHRONE_READY':
      draft.map.isochroneRings = action.payload;
      break;
    case 'ADDRESS_CHANGED':
      draft.map.currentLocation = action.payload;
      break;
    case 'FILTER_INPUT_CHANGED':
      draft.filter = action.payload;
      break;
    case 'FETCH_DELIVERY_POINTS_SUCCESS':
      draft.deliveryPoints = action.payload;
      break;
    case 'START_LOADING':
      draft.isLoading = true;
      break;
    case 'LOAD_SUCCESSFUL':
      draft.isLoading = false;
      break;
  }
});
