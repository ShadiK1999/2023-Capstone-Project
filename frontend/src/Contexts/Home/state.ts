import { Dispatch, createContext } from 'react';
import { HomeAction } from './reducer';
import { DisplayPoint } from '../../Models/DeliveryPoint';

type IsochroneRings = {
  inner: google.maps.LatLngLiteral[];
  middle: google.maps.LatLngLiteral[];
  outer: google.maps.LatLngLiteral[];
};

type Map = {
  currentLocation: google.maps.LatLngLiteral;
  isochroneRings: IsochroneRings | null;
};

export type DayOfWeekIndex = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export type Filter = {
  type: 'distance' | 'duration';
  mode: 'WALKING' | 'DRIVING' | 'CYCLING';
  value: number;
  dayOfWeek: DayOfWeekIndex;
  startHour: number;
  startMin: number;
  endHour: number;
  endMin: number;
};

export type HomeState = {
  map: Map;
  deliveryPoints: DisplayPoint[] | null;
  isLoading: boolean;
  filter: Filter;
};

export const initialState: HomeState = {
  map: {
    currentLocation: {
      lat: -37.8124,
      lng: 144.9623,
    },
    isochroneRings: null,
  },
  deliveryPoints: null,
  isLoading: false,
  filter: {
    type: 'duration',
    mode: 'WALKING',
    value: 5,
    dayOfWeek: 0,
    startHour: 9,
    startMin: 0,
    endHour: 17,
    endMin: 0,
  },
};

export const HomeContext = createContext<HomeState>(initialState);
export const HomeDispatchContext = createContext<Dispatch<HomeAction>>(() => void 0);
