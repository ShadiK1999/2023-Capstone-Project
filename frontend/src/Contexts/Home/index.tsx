import { FC, ReactNode, useReducer } from 'react';
import { HomeContext, HomeDispatchContext, initialState } from './state';
import { homeReducer } from './reducer';

type ComponentProps = {
  children: ReactNode;
};

const HomeProvider: FC<ComponentProps> = ({ children }) => {
  const [state, dispatch] = useReducer(homeReducer, initialState);

  return (
    <HomeContext.Provider value={state}>
      <HomeDispatchContext.Provider value={dispatch}>{children}</HomeDispatchContext.Provider>
    </HomeContext.Provider>
  );
};

export { HomeProvider };
