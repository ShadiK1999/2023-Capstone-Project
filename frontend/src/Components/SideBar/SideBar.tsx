import { useContext } from 'react';
import * as styles from './SideBar.styles';
import { HomeContext } from '../../Contexts/Home/state';
import { PointPanel } from './PointPanel';
import { Loader } from '../Loader';

const SideBar = () => {
  const { deliveryPoints, isLoading } = useContext(HomeContext);

  return (
    <div className={styles.Container}>
      {isLoading ? (
        <Loader />
      ) : (
        <div className={styles.InnerContainer} data-testid="inner-container">
          {deliveryPoints ? (
            !deliveryPoints.length ? (
              <p>No Delivery points found, please search again with a higher range</p>
            ) : (
              deliveryPoints.map((pt, index) => <PointPanel key={index} point={pt}></PointPanel>)
            )
          ) : null}
        </div>
      )}
    </div>
  );
};

export { SideBar };
