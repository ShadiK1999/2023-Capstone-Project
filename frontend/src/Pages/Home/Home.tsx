import { Map, InputBar, SideBar } from '../../Components';
import { HomeProvider } from '../../Contexts';

import * as styles from './Home.styles';

const Home = () => {
  return (
    <HomeProvider>
      <PageInternal />
    </HomeProvider>
  );
};

const PageInternal = () => {
  return (
    <div className={styles.Container}>
      <InputBar />
      <div></div>
      <Map />
      <SideBar />
    </div>
  );
};

export { Home };
