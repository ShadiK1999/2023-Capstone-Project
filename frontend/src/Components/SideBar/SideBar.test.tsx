// eslint-disable-next-line import/named
import { screen, render } from '@testing-library/react';
import { SideBar } from './SideBar';
import { describe, it, expect, vi, afterEach, beforeEach } from 'vitest';
import { HomeContext, HomeState, initialState } from '../../Contexts/Home/state';
import { DisplayPoint } from '../../Models/DeliveryPoint';
import { PointPanel } from './PointPanel';

const mockedPoint1: DisplayPoint = {
  id: undefined,
  address: '1 elizabeth st',
  distanceText: '1 km',
  timeText: '15 minutes',
  location: {
    lat: 0,
    lng: 0,
  },
  hours: {
    1: {
      openingTime: { hour: 9, minute: 0 },
      closingTime: { hour: 17, minute: 0 },
    }, // Open on Monday from 9 o'clock to 17 o'clock
  },
};

const mockedPoint2: DisplayPoint = {
  id: undefined,
  address: '2 collins st',
  distanceText: '2 km',
  timeText: '30 minutes',
  location: {
    lat: 0,
    lng: 0,
  },
  hours: {}, // Closed on Monday
};

const mockedState: HomeState = {
  ...initialState,
  deliveryPoints: [mockedPoint1, mockedPoint2],
};

describe('Sidebar', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date(2023, 8, 4)); // Monday. Months in Node JS is 0 based from 0 - 11 so this date means 4/9/2023. Trippy, I know
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('Matches DOM Snapshot', () => {
    // Arrange
    const tree = render(
      <HomeContext.Provider value={mockedState}>
        <SideBar />
      </HomeContext.Provider>,
    );

    // Assert
    expect(tree).toMatchSnapshot();
  });

  it('Displays 2 points given 2 points in state', () => {
    // Arrange
    render(
      <HomeContext.Provider value={mockedState}>
        <SideBar />
      </HomeContext.Provider>,
    );

    // Assert
    expect(screen.getByTestId('inner-container').childNodes.length).toEqual(2);
    expect(screen.getAllByText('1 elizabeth st')).toHaveLength(1);
    expect(screen.getAllByText('2 collins st')).toHaveLength(1);
  });

  it('Displays not found given empty array in state', () => {
    // Arrange
    render(
      <HomeContext.Provider value={{ ...mockedState, deliveryPoints: [] }}>
        <SideBar />
      </HomeContext.Provider>,
    );

    // Assert
    expect(
      screen.getAllByText('No Delivery points found, please search again with a higher range'),
    ).toHaveLength(1);
  });

  it('Displays loading if status is loading in state', () => {
    // Arrange
    render(
      <HomeContext.Provider value={{ ...mockedState, deliveryPoints: [], isLoading: true }}>
        <SideBar />
      </HomeContext.Provider>,
    );

    // Assert
    expect(screen.queryByTestId('inner-container')).toBeNull();
  });
});

describe('Point panel', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date(2023, 8, 4)); // Monday. Months in Node JS is 0 based from 0 - 11 so this date means 4/9/2023. Trippy, I know
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('Matches DOM Snapshot', () => {
    // Arrange
    const tree = render(<PointPanel point={mockedPoint1} />);

    // Assert
    expect(tree).toMatchSnapshot();
  });

  it('Displays opening hours if open', () => {
    // Arrange
    render(<PointPanel point={mockedPoint1} />);

    // Assert
    expect(screen.queryAllByText('09:00 - 17:00')).toHaveLength(1);
  });

  it('Displays closed if not open', () => {
    // Arrange
    render(<PointPanel point={mockedPoint2} />);

    // Assert
    expect(screen.queryAllByText('Closed')).toHaveLength(1);
  });
});
