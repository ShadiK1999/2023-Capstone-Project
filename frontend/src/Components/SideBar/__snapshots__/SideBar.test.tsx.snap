// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Point panel > Matches DOM Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="f3a03t2"
      >
        <p
          class="f1fepghe"
        >
          Location
        </p>
        <p
          class="f12a2168"
        >
          1 elizabeth st
        </p>
        <div
          class="f1kyo8ql"
        >
          <div
            class="fmcm93l"
          >
            <img
              src="/src/Components/SideBar/Assets/Clock.svg"
            />
            <p>
              09:00 - 17:00
            </p>
          </div>
          <div
            class="fmcm93l"
          >
            <img
              src="/src/Components/SideBar/Assets/Walking.svg"
            />
            <p>
              15 minutes
            </p>
          </div>
          <div
            class="fmcm93l"
          >
            <img
              src="/src/Components/SideBar/Assets/Target.svg"
            />
            <p>
              1 km
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="f3a03t2"
    >
      <p
        class="f1fepghe"
      >
        Location
      </p>
      <p
        class="f12a2168"
      >
        1 elizabeth st
      </p>
      <div
        class="f1kyo8ql"
      >
        <div
          class="fmcm93l"
        >
          <img
            src="/src/Components/SideBar/Assets/Clock.svg"
          />
          <p>
            09:00 - 17:00
          </p>
        </div>
        <div
          class="fmcm93l"
        >
          <img
            src="/src/Components/SideBar/Assets/Walking.svg"
          />
          <p>
            15 minutes
          </p>
        </div>
        <div
          class="fmcm93l"
        >
          <img
            src="/src/Components/SideBar/Assets/Target.svg"
          />
          <p>
            1 km
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Sidebar > Matches DOM Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="f1o03ftt"
      >
        <div
          class="ftnx03m"
          data-testid="inner-container"
        >
          <div
            class="f3a03t2"
          >
            <p
              class="f1fepghe"
            >
              Location
            </p>
            <p
              class="f12a2168"
            >
              1 elizabeth st
            </p>
            <div
              class="f1kyo8ql"
            >
              <div
                class="fmcm93l"
              >
                <img
                  src="/src/Components/SideBar/Assets/Clock.svg"
                />
                <p>
                  09:00 - 17:00
                </p>
              </div>
              <div
                class="fmcm93l"
              >
                <img
                  src="/src/Components/SideBar/Assets/Walking.svg"
                />
                <p>
                  15 minutes
                </p>
              </div>
              <div
                class="fmcm93l"
              >
                <img
                  src="/src/Components/SideBar/Assets/Target.svg"
                />
                <p>
                  1 km
                </p>
              </div>
            </div>
          </div>
          <div
            class="f3a03t2"
          >
            <p
              class="f1fepghe"
            >
              Location
            </p>
            <p
              class="f12a2168"
            >
              2 collins st
            </p>
            <div
              class="f1kyo8ql"
            >
              <div
                class="fmcm93l"
              >
                <img
                  src="/src/Components/SideBar/Assets/Clock.svg"
                />
                <p>
                  Closed
                </p>
              </div>
              <div
                class="fmcm93l"
              >
                <img
                  src="/src/Components/SideBar/Assets/Walking.svg"
                />
                <p>
                  30 minutes
                </p>
              </div>
              <div
                class="fmcm93l"
              >
                <img
                  src="/src/Components/SideBar/Assets/Target.svg"
                />
                <p>
                  2 km
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="f1o03ftt"
    >
      <div
        class="ftnx03m"
        data-testid="inner-container"
      >
        <div
          class="f3a03t2"
        >
          <p
            class="f1fepghe"
          >
            Location
          </p>
          <p
            class="f12a2168"
          >
            1 elizabeth st
          </p>
          <div
            class="f1kyo8ql"
          >
            <div
              class="fmcm93l"
            >
              <img
                src="/src/Components/SideBar/Assets/Clock.svg"
              />
              <p>
                09:00 - 17:00
              </p>
            </div>
            <div
              class="fmcm93l"
            >
              <img
                src="/src/Components/SideBar/Assets/Walking.svg"
              />
              <p>
                15 minutes
              </p>
            </div>
            <div
              class="fmcm93l"
            >
              <img
                src="/src/Components/SideBar/Assets/Target.svg"
              />
              <p>
                1 km
              </p>
            </div>
          </div>
        </div>
        <div
          class="f3a03t2"
        >
          <p
            class="f1fepghe"
          >
            Location
          </p>
          <p
            class="f12a2168"
          >
            2 collins st
          </p>
          <div
            class="f1kyo8ql"
          >
            <div
              class="fmcm93l"
            >
              <img
                src="/src/Components/SideBar/Assets/Clock.svg"
              />
              <p>
                Closed
              </p>
            </div>
            <div
              class="fmcm93l"
            >
              <img
                src="/src/Components/SideBar/Assets/Walking.svg"
              />
              <p>
                30 minutes
              </p>
            </div>
            <div
              class="fmcm93l"
            >
              <img
                src="/src/Components/SideBar/Assets/Target.svg"
              />
              <p>
                2 km
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
